README.md
# mwser-ucnext

mwser-environment setup for ucnext. Uses the chef library/wrapper pattern and includes [universityofcalifornia/chef-ucnext](https://github.com/univeristyofcalifornia/chef-ucnext).

## Supported Platforms

CentOS6

## Usage

### mwser-ucnext::default

Include `mwser-ucnext` in your node's `run_list`.

## License and Authors

Author:: Steve Nolen (technolengy@gmail.com)


CHANGELOG.md
# 0.1.0

Initial release of mwser-ucnext


recipes/_bridge.rb
# adds shib-oauth2-bridge

# additional packages/repos needed.
yum_repository 'remi' do
  description 'Les RPM de Remi - Repository'
  mirrorlist 'http://rpms.famillecollet.com/enterprise/6/remi/mirror'
  gpgkey 'http://rpms.famillecollet.com/RPM-GPG-KEY-remi'
  action :create
end

yum_repository 'remi-php55' do
  description 'Les RPM de Remi PHP55 - Repository'
  mirrorlist 'http://rpms.famillecollet.com/enterprise/6/php55/mirror'
  gpgkey 'http://rpms.famillecollet.com/RPM-GPG-KEY-remi'
  action :create
end

%w(git php php-mcrypt php-mysql php-mbstring).each do |pkg|
  package pkg
end

# shib config
include_recipe 'shib-oauth2-bridge::shibd'
include_recipe 'shib-oauth2-bridge::shib-ds'

fqdn = node['fqdn'] # quick variable accessor

# db config
db_root_obj = ChefVault::Item.load("passwords", "db_root")
db_root = db_root_obj[fqdn]
mysql_connection = {
  :host => '127.0.0.1',
  :port => 3306,
  :username => 'root',
  :password => db_root
}

db_bridge_obj = ChefVault::Item.load('passwords', 'bridge')
db_bridge = db_bridge_obj[fqdn]

mysql_database 'bridge' do
  connection mysql_connection
  action :create
end
mysql_database_user 'bridge' do
  connection mysql_connection
  password db_bridge
  database_name 'bridge'
  action [:create,:grant]
end

# shib keys
sp_ssl = ChefVault::Item.load('shibboleth', fqdn) # gets ssl cert from chef-vault
file '/etc/shibboleth/sp-cert.pem' do
  owner 'shibd'
  group 'shibd'
  mode '0777'
  content sp_ssl['cert']
  notifies :reload, 'service[shibd]', :delayed
end
file '/etc/shibboleth/sp-key.pem' do
  owner 'shibd'
  group 'shibd'
  mode '0600'
  content sp_ssl['key']
  notifies :reload, 'service[shibd]', :delayed
end

# shib bridge setup
bridge_secrets = ChefVault::Item.load('secrets', 'oauth2')
shib_oauth2_bridge 'default' do
  db_user 'bridge'
  db_name 'bridge'
  hostname 'ucnext.org'
  db_password db_bridge
  clients [
    {id: 'next', name: 'next', secret: bridge_secrets['next'], redirect_uri: 'https://ucnext.org/auth/oauth2/shibboleth'},
    {id: 'staging_next', name: 'staging_next', secret: bridge_secrets['staging_next'], redirect_uri: 'https://staging.ucnext.org/auth/oauth2/shibboleth'}
  ]
end


recipes/default.rb
#
# Cookbook Name:: mwser-ucnext
# Recipe:: default
#
# Copyright (C) 2015 UC Regents
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# require chef-vault
chef_gem 'chef-vault'
require 'chef-vault'

# some basic package deps. only tested on rhel family.
package 'git'

fqdn = if node['fqdn'] == 'ucnext.org'
         'ucnext.org'
       else
         'staging.ucnext.org'
       end

# install mysql
db_root_obj = ChefVault::Item.load("passwords", "db_root")
db_root = db_root_obj[node['fqdn']]
db_next_obj = ChefVault::Item.load("passwords", "next")
db_next = db_next_obj[fqdn]
mysql_service 'default' do
  port '3306'
  version '5.6'
  initial_root_password db_root
  action [:create, :start]
end
mysql_config 'utf8mb4' do
  source 'utf8mb4.erb'
  instance 'default'
  notifies :restart, 'mysql_service[default]'
  action :create
end

mysql_connection = {
  :host => '127.0.0.1',
  :port => 3306,
  :username => 'root',
  :password => db_root
}

# set up ucnext db
mysql2_chef_gem 'default'
mysql_database 'next' do
  connection mysql_connection
  action :create
end
mysql_database_user 'next' do
  connection mysql_connection
  password db_next
  database_name 'next'
  action [:create,:grant]
end

# a few case-y things based on hostname
case fqdn
when 'ucnext.org'
  app_name = 'prod' # name of ucnext service
  shib_client = 'next'
  bridge_enabled = true
  app_revision = '1.0.41'
  rails_env = 'production'
when 'staging.ucnext.org'
  app_name = 'staging'
  shib_client = 'staging_next'
  bridge_enabled = false
  app_revision = 'master'
  rails_env = 'staging'
end

# install nginx
node.set['nginx']['default_site_enabled'] = false
node.set['nginx']['install_method'] = 'package'
include_recipe 'nginx::repo'
include_recipe 'nginx'

directory '/etc/ssl/private' do
  recursive true
end

# add SSL certs to box
ssl_key_cert = ChefVault::Item.load('ssl', fqdn) # gets ssl cert from chef-vault
file "/etc/ssl/certs/#{fqdn}.crt" do
  owner 'root'
  group 'root'
  mode '0777'
  content ssl_key_cert['cert']
  notifies :reload, 'service[nginx]', :delayed
end
file "/etc/ssl/private/#{fqdn}.key" do
  owner 'root'
  group 'root'
  mode '0600'
  content ssl_key_cert['key']
  notifies :reload, 'service[nginx]', :delayed
end

# nginx conf
template '/etc/nginx/sites-available/ucnext' do
  source 'ucnext.conf.erb'
  mode '0775'
  action :create
  variables(
    port: 3000,
    path: '/var/www/', # not used.
    bridge_enabled: bridge_enabled,
    fqdn: fqdn
  )
  notifies :reload, 'service[nginx]', :delayed
end
nginx_site 'ucnext' do
  action :enable
end

# install ruby with rbenv, npm, git
node.default['rbenv']['rubies'] = ['2.2.3']
include_recipe 'ruby_build'
include_recipe 'ruby_rbenv::system'
include_recipe 'nodejs::npm'
rbenv_global '2.2.3'
rbenv_gem 'bundle'

rails_secrets = ChefVault::Item.load('secrets', 'rails_secret_tokens')
smtp = ChefVault::Item.load('smtp', 'ucnext.org')
bridge_secrets = ChefVault::Item.load('secrets', 'oauth2')

# set up ucnext!
ucnext app_name do
  revision app_revision
  port 3000
  secret rails_secrets[fqdn]
  db_password db_next
  deploy_path '/var/next'
  bundler_path '/usr/local/rbenv/shims'
  smtp_host smtp['host']
  smtp_username smtp['username']
  smtp_password smtp['password']
  shib_client_name shib_client
  shib_secret bridge_secrets[shib_client]
  rails_env rails_env
  # assumes es_host is localhost!
end
